File: backend/simulation/signals.py
# backend/simulation/signals.py
# SwiftAid — Traffic signal management
#
# Signal states (in order):
#   RED  →  PREPARE  →  GREEN  →  (ambulance passes)  →  RECOVERY  →  RED/GREEN
#
# PREPARE is the small delay before a signal goes green after being pre-empted.
# RECOVERY is the amber cooldown after ambulance clears the intersection.
# This makes the demo look more realistic than instant flips.

import random
from utils.constants import CONFIG


# All possible signal states
STATES = ['red', 'prepare', 'green', 'amber', 'recovery']


class SignalManager:
    """
    Manages all traffic signals on the grid.
    Signals cycle normally unless overridden by the AI for an ambulance.
    """

    def __init__(self, grid_size: int):
        self.G = grid_size
        self.signals:       dict = {}
        self.override_count: int = 0
        self.stops_avoided:  int = 0   # metric: how many red signals were bypassed
        self.reset()

    def reset(self):
        """Reinitialise all signals to random red/green states."""
        self.signals        = {}
        self.override_count = 0
        self.stops_avoided  = 0
        for r in range(1, self.G):
            for c in range(1, self.G):
                self.signals[f'{r},{c}'] = {
                    'state':         'green' if random.random() < 0.5 else 'red',
                    'timer':         random.randint(0, CONFIG['SIGNAL_CYCLE']),
                    'override':      False,
                    'prepare_timer': 0,      # countdown for PREPARE phase
                    'recover_timer': 0,      # countdown for RECOVERY phase
                    'overridden_by': None,   # ambulance id that owns this override
                }

    # ── Per-tick update ──────────────────────────────────────────────
    def tick(self):
        """
        Advance all signals by one tick.
        Overridden signals are skipped — the AI owns them.
        """
        for s in self.signals.values():

            # AI-controlled: handle PREPARE phase
            if s['override']:
                if s['state'] == 'prepare':
                    s['prepare_timer'] -= 1
                    if s['prepare_timer'] <= 0:
                        s['state'] = 'green'   # finally go green
                continue

            # Recovery phase after ambulance passed
            if s['state'] == 'recovery':
                s['recover_timer'] -= 1
                if s['recover_timer'] <= 0:
                    s['state'] = 'green' if random.random() < 0.5 else 'red'
                    s['timer']  = CONFIG['SIGNAL_CYCLE'] + random.randint(0, 20)
                continue

            # Normal cycle: red ↔ green
            s['timer'] -= 1
            if s['timer'] <= 0:
                s['state'] = 'red' if s['state'] == 'green' else 'green'
                s['timer']  = CONFIG['SIGNAL_CYCLE'] + random.randint(0, 30)

    # ── Pre-emption (AI override) ────────────────────────────────────
    def preempt(self, r: int, c: int, amb_id: int) -> bool:
        """
        Pre-empt a signal for an ambulance.
        Goes through PREPARE state first (small delay) then GREEN.
        Returns True if this is a new override.
        """
        k = f'{r},{c}'
        s = self.signals.get(k)
        if not s:
            return False
        if s['override'] and s['overridden_by'] == amb_id:
            return False   # already handling this one

        was_red = (s['state'] == 'red')

        s['override']      = True
        s['overridden_by'] = amb_id
        self.override_count += 1

        # If currently red: go through PREPARE phase first
        if was_red:
            s['state']         = 'prepare'
            s['prepare_timer'] = CONFIG['SIGNAL_PREPARE_TICKS']
            if s.get('state') == 'red':
                self.stops_avoided += 1   # would have stopped here
        else:
            s['state'] = 'green'   # already green, keep it

        return True

    # ── Release (ambulance passed) ───────────────────────────────────
    def release(self, r: int, c: int, amb_id: int):
        """
        Release signal after ambulance clears the intersection.
        Signal enters RECOVERY (amber) phase before resuming normal cycle.
        """
        k = f'{r},{c}'
        s = self.signals.get(k)
        if not s or not s['override'] or s['overridden_by'] != amb_id:
            return
        s['override']      = False
        s['overridden_by'] = None
        s['state']         = 'recovery'
        s['recover_timer'] = CONFIG['SIGNAL_RECOVER']

    # ── Scenario chaos ───────────────────────────────────────────────
    def chaos(self):
        """Randomise signal timers — used by peak/conflict scenarios."""
        for s in self.signals.values():
            if random.random() < 0.45:
                s['state'] = 'red'   # bias toward red for harder demo
                s['timer']  = random.randint(3, 25)

    # ── Queries ──────────────────────────────────────────────────────
    def get_color(self, r: int, c: int) -> str:
        s = self.signals.get(f'{r},{c}')
        if not s: return '#888'
        return {
            'green':    '#1a9e4a',
            'red':      '#d63b3b',
            'prepare':  '#f59e0b',  # yellow — about to go green
            'recovery': '#e8981f',  # amber  — cooling down
        }.get(s['state'], '#888')

    def snapshot(self) -> dict:
        """Return signal states for API response."""
        return {
            k: {'state': s['state'], 'override': s['override']}
            for k, s in self.signals.items()
        }

File: frontend/static/js/signals.js
// simulation/signals.js — Traffic signal logic and override system

class SignalManager {
  constructor(gridSize) {
    this.G = gridSize;
    this.signals = {};
    this.overrideCount = 0;
    this.init();
  }

  init() {
    for (let r = 1; r < this.G; r++) {
      for (let c = 1; c < this.G; c++) {
        this.signals[`${r},${c}`] = {
          state: Math.random() < 0.5 ? 'green' : 'red',
          timer: Math.floor(Math.random() * CONFIG.SIGNAL_CYCLE),
          override: false,
          recovering: false,
          recoverTimer: 0,
          overriddenBy: null, // ambulance id
        };
      }
    }
    this.overrideCount = 0;
  }

  reset() { this.init(); }

  key(r, c) { return `${r},${c}`; }

  get(r, c) { return this.signals[this.key(r, c)]; }

  // Called every tick — advances timers and cycles normal signals
  tick(aiEnabled) {
    for (const k in this.signals) {
      const s = this.signals[k];
      if (s.override) continue;

      if (s.recovering) {
        s.recoverTimer--;
        if (s.recoverTimer <= 0) {
          s.recovering = false;
          s.state = Math.random() < 0.5 ? 'green' : 'red';
          s.timer = CONFIG.SIGNAL_CYCLE;
        }
        continue;
      }

      s.timer--;
      if (s.timer <= 0) {
        s.state = s.state === 'green' ? 'red' : 'green';
        s.timer = CONFIG.SIGNAL_CYCLE + Math.floor(Math.random() * 30);
      }
    }
  }

  // Pre-empt a signal (turn green for ambulance)
  preempt(r, c, ambId) {
    const s = this.get(r, c);
    if (!s) return false;
    if (s.override && s.overriddenBy === ambId) return false; // already done
    s.override = true;
    s.state = 'green';
    s.overriddenBy = ambId;
    this.overrideCount++;
    return true;
  }

  // Release a signal after ambulance passes
  release(r, c, ambId) {
    const s = this.get(r, c);
    if (!s || !s.override || s.overriddenBy !== ambId) return;
    s.override = false;
    s.recovering = true;
    s.state = 'amber';
    s.recoverTimer = CONFIG.SIGNAL_RECOVER;
    s.overriddenBy = null;
  }

  // Resolve conflict: two ambulances at same intersection
  resolveConflict(r, c, ambA, ambB) {
    // Higher priority wins; tie-break by id
    const winner = ambA.priority <= ambB.priority ? ambA : ambB;
    const loser  = winner === ambA ? ambB : ambA;
    this.preempt(r, c, winner.id);
    // Loser must wait — signal it red briefly
    loser.waiting = true;
    loser.waitTimer = 30;
    return winner;
  }

  getColor(r, c) {
    const s = this.get(r, c);
    if (!s) return null;
    if (s.state === 'green') return '#7adf3a';
    if (s.state === 'amber') return '#EF9F27';
    return '#E24B4A';
  }

  // Inject chaos (scramble some signals)
  chaos() {
    for (const k in this.signals) {
      if (Math.random() < 0.4) {
        this.signals[k].state = Math.random() < 0.5 ? 'green' : 'red';
        this.signals[k].timer = Math.floor(Math.random() * 30);
      }
    }
  }
}
